src/dbnode/storage/block/lease.go

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package block
+
+import (
+	"errors"
+	"sync"
+)
+
+var (
+	errLeaserAlreadyRegistered = errors.New("leaser already registered")
+	errAddLeaseLeaserNotRegistered = errors.New("add lease error as leaser not registered")
+)
+
+type leaseManager struct {
+	sync.Mutex
+	leasers []Leaser
+	verifier LeaseVerifier
+}
+
+// NewLeaseManager creates a new lease manager with a provided
+// lease verifier (to ensure leases are valid when made).
+func NewLeaseManager(verifier LeaseVerifier) LeaseManager {
+	return &leaseManager{
+		verifier: verifier,
+	}
+}
+
+func (m *leaseManager) RegisterLeaser(leaser Leaser) error {
+	m.Lock()
+	defer m.Unlock()
+
+	for _, l := range m.leasers {
+		if l == leaser {
+			return errLeaserAlreadyRegistered
+		}
+	}
+
+	m.leasers = append(m.leasers, leaser)
+
+	return nil
+}
+
+func (m *leaseManager) OpenLease( 
```

Comment:
I'm just now thinking about this but not sure how we're gonna handle the bootstrap case where we call `CacheShardIndices`

That will have to open up a bunch of seekers, but the node is not bootstrapped yet so we won't have the correct state for everything yet so the verifier won't be able to do much....do we just have the verifier skip verification if the node is not bootstrapped yet?

Reply:
I guess the other alternative would be to pull move the CacheShardIndices call to somewhere after all the shard/namespaces are bootstrapped but before we mark the actual database bootstrapped instead of doing it in the bootstrappers themselves

Reply:
As discussed, moving CacheShardIndices to after shards are bootstrapped (but before namespace is bootstrapped IMO, because that's when reads start actually being available) is a good idea.